src/scib_metrics/benchmark/_core.py

```diff
@@ -0,0 +1,338 @@
+        batch_key: str,
+        label_key: str,
+        embedding_obsm_keys: List[str],
+        bio_conservation_metrics: Optional[BioConservation] = None,
+        batch_correction_metrics: Optional[BatchCorrection] = None,
+        pre_integrated_embedding_obsm_key: Optional[str] = None,
+        n_jobs: int = 1,
+    ):
+        self._adata = adata
+        self._embedding_obsm_keys = embedding_obsm_keys
+        self._pre_integrated_embedding_obsm_key = pre_integrated_embedding_obsm_key
+        self._bio_conservation_metrics = bio_conservation_metrics if bio_conservation_metrics else BioConservation()
+        self._batch_correction_metrics = batch_correction_metrics if batch_correction_metrics else BatchCorrection()
+        self._results = pd.DataFrame(columns=list(self._embedding_obsm_keys) + [_METRIC_TYPE])
+        self._emb_adatas = {}
+        self._neighbor_values = (15, 50, 90)
+        self._prepared = False
+        self._batch_key = batch_key
+        self._label_key = label_key
+        self._n_jobs = n_jobs
+
+        self._metric_collection_dict = {
+            "Bio conservation": self._bio_conservation_metrics,
+            "Batch correction": self._batch_correction_metrics,
+        }
+
+    def prepare(self) -> None:
+        """Prepare the data for benchmarking."""
+        # Compute PCA
+        if self._pre_integrated_embedding_obsm_key is None:
+            # This is how scib does it
+            # https://github.com/theislab/scib/blob/896f689e5fe8c57502cb012af06bed1a9b2b61d2/scib/metrics/pcr.py#L197
+            sc.tl.pca(self._adata, use_highly_variable=False)
+            self._pre_integrated_embedding_obsm_key = "X_pca"
+
+        for emb_key in self._embedding_obsm_keys:
+            self._emb_adatas[emb_key] = AnnData(self._adata.obsm[emb_key], obs=self._adata.obs)
+            self._emb_adatas[emb_key].obs[_BATCH] = np.asarray(self._adata.obs[self._batch_key].values)
+            self._emb_adatas[emb_key].obs[_LABELS] = np.asarray(self._adata.obs[self._label_key].values)
+            self._emb_adatas[emb_key].obsm[_X_PRE] = self._adata.obsm[self._pre_integrated_embedding_obsm_key]
+
+        # Compute neighbors
+        for ad in tqdm(self._emb_adatas.values(), desc="Computing neighbors"):
+            knn_search_index = NNDescent(
+                ad.X,
+                n_neighbors=max(self._neighbor_values),
+                random_state=0,
+                low_memory=True,
+                n_jobs=self._n_jobs,
+            )
+            indices, distances = knn_search_index.neighbor_graph
+            for n in self._neighbor_values:
+                sp_distances, sp_conns = sc.neighbors._compute_connectivities_umap(
+                    indices[:, :n], distances[:, :n], ad.n_obs, n_neighbors=n
+                )
+                ad.obsp[f"{n}_connectivities"] = sp_conns
+                ad.obsp[f"{n}_distances"] = sp_distances
+
+        self._prepared = True
+
+    def benchmark(self) -> None:
```

Comment:
should this fn avoid incomplete overrides of results? It seems like there are few to no cases where it should be called multiple times but it would just redo computation here. Specifically confusing in cases where somebody calls benchmark againintentionally after tweaking metrics they want to run, this would result in a mishmash of old and new results.

Reply:
can use frozen in the metrics dataclasses so that it wouldn't be possible to run benchmark with different metrics